internal/pkg/source/source_test.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"testing"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/analysistest"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+	"golang.org/x/tools/go/ssa"
+)
+
+type testConfig struct {
+	propagatorsPattern string
+	fieldsPattern      string
+	sanitizerPattern   string
+}
+
+func (c *testConfig) IsSanitizer(call *ssa.Call) bool {
+	match, _ := regexp.MatchString(c.sanitizerPattern, call.String())
+	return match
+}
+
+func (c *testConfig) IsPropagator(call *ssa.Call) bool {
+	match, _ := regexp.MatchString(c.propagatorsPattern, call.String())
+	return match
+}
+
+func (c *testConfig) IsSourceFieldAddr(field *ssa.FieldAddr) bool {
+	match, _ := regexp.MatchString(c.fieldsPattern, field.String())
+	return match
+}
+
+type analyzerResult struct {
+	allocations []*ssa.Alloc
+	calls       []*ssa.Call
+	fieldAddr   []*ssa.FieldAddr
+	store       []*ssa.Store
+}
+
+var testAnalyzer = &analysis.Analyzer{
+	Name:       "source",
+	Run:        run,
+	Doc:        "test harness for the logic related to sources",
+	Requires:   []*analysis.Analyzer{buildssa.Analyzer},
+	ResultType: reflect.TypeOf(analyzerResult{}),
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	in := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+	var result analyzerResult
+	for _, fn := range in.SrcFuncs {
+		fmt.Printf("Func: %v\n", fn.Name())
+		for _, b := range fn.Blocks {
+			for _, i := range b.Instrs {
+				fmt.Printf("%v : %T\n", i, i)
+
+				switch v := i.(type) {
+				case *ssa.Alloc:
+					result.allocations = append(result.allocations, v)
+				case *ssa.Call:
+					result.calls = append(result.calls, v)
+				case *ssa.FieldAddr:
+					result.fieldAddr = append(result.fieldAddr, v)
+				case *ssa.Store:
+					result.store = append(result.store, v)
+				}
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func TestSource(t *testing.T) {
+	dir := analysistest.TestData()
+	config := &testConfig{
+		propagatorsPattern: "propagator",
+		sanitizerPattern:   "sanitizer",
+		fieldsPattern:      "name",
+	}
+
+	testCases := []struct {
+		desc                         string
+		pattern                      string
+		config                       *testConfig
+		wantCallConnectionIdx        []int
+		wantFieldAccessConnectionIdx []int
+		wantStoreConnectionIdx       []int
```

Comment:
Testing based on index in the `ssa` result feels to me a bit brittle and resistant to future change.  See above comment about `// want` approach.

Reply:
See my previous [response](https://github.com/google/go-flow-levee/pull/11#discussion_r424108402).